File: src/apps/gymmanager-web/src/types/transaction.ts
export type TransactionType =
  | "MembershipFee"
  | "SalaryPayment"
  | "Rent"
  | "Utilities"
  | "Equipment"
  | "Wages"
  | "Expense"
  | "Refund"
  | "Other";

export type TransactionDirection = "Credit" | "Debit";

export type TransactionCategory =
  | "Revenue"
  | "OperatingExpense"
  | "CapitalExpense"
  | "Payroll"
  | "Refund";

export type PaymentMethod = "Cash" | "BankTransfer" | "Card" | "Online";

export interface TransactionDto {
  id: string;
  gymHouseId: string;
  transactionType: TransactionType;
  direction: TransactionDirection;
  amount: number;
  category: TransactionCategory;
  description: string;
  transactionDate: string;
  relatedEntityId: string | null;
  reversesTransactionId: string | null;
  reversedByTransactionId: string | null;
  approvedById: string | null;
  paymentMethod: PaymentMethod | null;
  externalReference: string | null;
  createdAt: string;
}

export interface PnLLineDto {
  category: TransactionCategory;
  totalAmount: number;
}

export interface PnLReportDto {
  gymHouseId: string | null;
  from: string;
  to: string;
  incomeLines: PnLLineDto[];
  expenseLines: PnLLineDto[];
  totalIncome: number;
  totalExpense: number;
  netProfit: number;
}

export interface RevenueMetricsDto {
  gymHouseId: string;
  from: string;
  to: string;
  mrr: number;
  churnRate: number;
  avgRevenuePerMember: number;
  totalRevenue: number;
  activeMembers: number;
  cancelledSubscriptions: number;
}

export interface RecordTransactionRequest {
  transactionType: TransactionType;
  direction: TransactionDirection;
  amount: number;
  category: TransactionCategory;
  description: string;
  transactionDate: string;
  relatedEntityId?: string;
  approvedById?: string;
  paymentMethod?: PaymentMethod;
  externalReference?: string;
}

export interface ReverseTransactionRequest {
  reason: string;
}

export interface TransactionFilters {
  type?: TransactionType;
  direction?: TransactionDirection;
  from?: string;
  to?: string;
}

File: src/core/GymManager.Domain/Entities/Transaction.cs
using GymManager.Domain.Common;
using GymManager.Domain.Enums;

namespace GymManager.Domain.Entities;

public sealed class Transaction : AuditableEntity
{
    public Guid GymHouseId { get; set; }
    public TransactionType TransactionType { get; set; }
    public TransactionDirection Direction { get; set; }
    public decimal Amount { get; set; }
    public TransactionCategory Category { get; set; }
    public string Description { get; set; } = string.Empty;
    public DateTime TransactionDate { get; set; }
    public Guid? RelatedEntityId { get; set; }
    public Guid? ReversesTransactionId { get; set; }
    public Guid? ReversedByTransactionId { get; set; }
    public Guid? ApprovedById { get; set; }
    public PaymentMethod? PaymentMethod { get; set; }
    public string? ExternalReference { get; set; }

    public GymHouse GymHouse { get; set; } = null!;
    public Transaction? ReversesTransaction { get; set; }
    public Transaction? ReversedByTransaction { get; set; }

    public static Transaction CreateReversal(Transaction original, string reason) =>
        new()
        {
            GymHouseId = original.GymHouseId,
            TransactionType = original.TransactionType,
            Direction = original.Direction == TransactionDirection.Credit
                ? TransactionDirection.Debit
                : TransactionDirection.Credit,
            Amount = original.Amount,
            Category = original.Category,
            Description = $"Reversal of transaction {original.Id}: {reason}",
            TransactionDate = DateTime.UtcNow,
            ReversesTransactionId = original.Id,
            ApprovedById = original.ApprovedById,
            PaymentMethod = original.PaymentMethod
        };
}
